### PR TITLE
misc: add advanced endpoint config guide

### PR DESCRIPTION
**Pull request by lucix-aws**
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: docs/howto/configuring/endpoints.md [hunk 1]
```diff
@@ -1,26 +1,164 @@
...
+
+In the latter, you are specifying in absolute terms that requests should be made against the given example endpoint.
+
+**While these two settings are not mutually exclusive, in general, you can expect to only need to modify one of them
+depending on your use case. As a general SDK user, you will most often be making endpoint customizations
+through `endpointUrl`.**
+
+## A note about S3
+
+S3 is a depth-rich service with many of its features modeled through endpoint-related customizations, such as bucket
```
Comment (lauzadis) [inline docs/howto/configuring/endpoints.md:96]: nit: depth-rich -> feature-rich, or "complex service"

File: docs/howto/configuring/endpoints.md [hunk 2]
```diff
@@ -1,26 +1,164 @@
...
+virtual hosting, where the bucket name is inserted into the endpoint hostname for requests.
+
+Because of this, it is generally not recommended to replace the `EndpointProvider` implementation in S3. If you need to
+extend its resolution behavior (such as for accessing a preview feature), you will most likely need to wrap the default
```
Comment (lauzadis) [inline docs/howto/configuring/endpoints.md:100]: nit: preview features are an internal concept, external users likely won't do this

maybe change to "such as sending requests to a local development stack"

**Review by lauzadis — APPROVED** [review]

File: docs/howto/configuring/endpoints.md [hunk 3]
```diff
@@ -1,26 +1,164 @@
...
+
+### `EndpointProvider` parameters
+
+Each service takes a specific set of inputs which are passed to its resolution function, defined in each service client
+package as `EndpointParameters`.
+
+Every service includes the following base parameters, which are used to facilitate general endpoint resolution within
+AWS:
+
+* `region`
```
Comment (aajtodd) [inline docs/howto/configuring/endpoints.md:56]: nit: Could format this in a table

**Review by aajtodd — APPROVED** [review]

File: codegen/smithy-aws-kotlin-codegen/src/main/kotlin/aws/sdk/kotlin/codegen/AwsServiceConfigIntegration.kt [hunk 4]
```diff
@@ -76,9 +77,15 @@ class AwsServiceConfigIntegration : KotlinIntegration {
 
         val EndpointUrlProp = ConfigProperty {
             name = "endpointUrl"
-            symbol = RuntimeTypes.Core.Net.Url.toBuilder().boxed().build()
+            symbol = RuntimeTypes.Core.Net.Url.asNullable()
             documentation = """
-                A custom endpoint to use when making requests.
+                A custom endpoint to use when making requests. **This is an advanced config option.**
```
Comment (aajtodd) [inline codegen/smithy-aws-kotlin-codegen/src/main/kotlin/aws/sdk/kotlin/codegen/AwsServiceConfigIntegration.kt:82]: **Suggestion**: I don't think we want to use terms like "in practice" or "most likely". We should be more precise. What about the following?

> A custom endpoint to route requests to. The endpoint set here is passed to the configured [endpointProvider], which may inspect and modify it as needed. 
> 
> Setting a custom endpointUrl should generally be preferred to overriding the [endpointProvider] and is the recommended way to route requests to development or preview instances of a service.
>
> **This is an advanced configuration option.**

**Pull request merged**